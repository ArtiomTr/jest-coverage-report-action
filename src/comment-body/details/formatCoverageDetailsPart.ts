import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { ParsedCoverageDetails } from '../../collect-coverage/parseCoverageDetails';
import { details } from '../strings.json';
import { createMarkdownSpoiler } from '../utils/createMarkdownSpoiler';
import { formatHeadingAndTable } from '../utils/formatHeadingAndTable';

export type DetailsFormatOptions = {
    summary: string;
    heading: string;
};

export const formatCoverageDetailsPart = (
    formatOptions: DetailsFormatOptions,
    headDetails: ParsedCoverageDetails,
    baseDetails?: ParsedCoverageDetails
): string | undefined => {
    const { summary, heading } = formatOptions;

    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename]
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: formatHeadingAndTable(
                heading,
                markdownTable([details.columnHeaders, ...tableContent])
            ),
            summary,
        });
    }

    return undefined;
};
