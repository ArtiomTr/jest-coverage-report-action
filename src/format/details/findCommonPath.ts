import { dirname } from 'path';

/**
 * Find a common path from a list of filepaths.
 */
export const findCommonPath = (filepaths: string[]): string => {
    let commonRoot = '';
    if (filepaths.length) {
        // If the paths are sorted, any prefix common to all paths will be common to the sorted first and last strings.
        const sortedPaths = [...filepaths].sort();
        const first = sortedPaths[0];
        const last = sortedPaths[sortedPaths.length - 1];

        const len = Math.min(first.length, last.length);
        for (let i = 0; i < len; i++) {
            const ch1 = first[i];
            const ch2 = last[i];
            if (ch1 == ch2) {
                commonRoot += ch1;
            } else {
                break;
            }
        }

        // If it doesn't appear to be a directory (partial filename), get the parent directory
        if (commonRoot.length && commonRoot[commonRoot.length - 1] !== '/') {
            commonRoot = dirname(commonRoot) + '/';
        }
    }
    return commonRoot;
};
