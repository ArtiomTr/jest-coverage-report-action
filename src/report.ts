import core from '@actions/core';
import { ZodError } from 'zod';

import { getCurrentBranch } from './git';
import { getOptions } from './options';

export interface InvalidOptionsReport {
    type: 'invalid-options';
    error: ZodError;
}

export type CovbotReport = InvalidOptionsReport;

export const getReport = async (): Promise<CovbotReport> => {
    core.debug('[stage]: getting options');
    const options = getOptions();

    /**
     * Report error if failed to parse options.
     * There is nothing more we could do.
     */
    if (!options.success) {
        return {
            type: 'invalid-options',
            error: options.error,
        };
    }

    /**
     * Try to get current branch on current repository.
     * This is needed just for consistency - when action ends work, return to initial branch.
     * It will automatically report warning, if getting this branch failed.
     */
    core.debug('[stage]: getting current branch');
    const currentBranch = await getCurrentBranch();

    /**
     * If it is pull request, or pull request number is present, get PR info
     */
    const pullRequest = getPullRequestInfo();
};
