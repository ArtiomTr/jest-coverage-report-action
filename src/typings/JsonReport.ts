export type JsonReport = {
    numFailedTestSuites: number;
    numFailedTests: number;
    numPassedTestSuites: number;
    numPassedTests: number;
    numPendingTestSuites: number;
    numPendingTests: number;
    numRuntimeErrorTestSuites: number;
    numTodoTests: number;
    numTotalTestSuites: number;
    numTotalTests: number;
    openHandles?: unknown[];
    snapshot: Snapshot;
    startTime: number;
    success: boolean;
    testResults?: TestResult[];
    wasInterrupted: boolean;
    coverageMap: CoverageMap;
};

export type Snapshot = {
    added: number;
    didUpdate: boolean;
    failure: boolean;
    filesAdded: number;
    filesRemoved: number;
    filesRemovedList?: unknown[];
    filesUnmatched: number;
    filesUpdated: number;
    matched: number;
    total: number;
    unchecked: number;
    uncheckedKeysByFile?: unknown[];
    unmatched: number;
    updated: number;
};

export type TestResult = {
    assertionResults?: AssertionResult[];
    endTime: number;
    message: string;
    name: string;
    startTime: number;
    status: string;
    summary: string;
};

export type AssertionResult = {
    ancestorTitles?: string[];
    failureMessages?: string[];
    fullName: string;
    location: Location;
    status: string;
    title: string;
};

export type Location = {
    column?: number;
    line: number;
};

export type Range = {
    start?: Location;
    end?: Location;
};

export type CoverageMap = Record<string, FileCoverage>;

export type FileCoverage = {
    path: string;
    statementMap: StatementMap;
    fnMap: FunctionMap;
    branchMap: BranchMap;
    s: HitMap;
    f: HitMap;
    b: ArrayHitMap;
};

export type StatementMap = Record<number, StatementCoverage>;

export type StatementCoverage = {
    start: Location;
    end: Location;
};

export type FunctionMap = Record<number, FunctionCoverage>;

export type FunctionCoverage = {
    name: string;
    decl: Range;
    loc: Range;
};

export type BranchMap = Record<number, BranchCoverage>;

export type BranchCoverage = {
    loc: Range;
    type: string;
    locations?: Range[];
};

export type HitMap = Record<number, number>;

export type ArrayHitMap = Record<number, number[]>;
