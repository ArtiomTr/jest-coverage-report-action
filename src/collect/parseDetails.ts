import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from './counters';
import { findCommonPath } from './findCommonPath';
import { getPercents } from './getPercents';
import { CoverageDetailsMap } from '../typings/Coverage';
import { JsonReport } from '../typings/JsonReport';

export const parseDetails = (jsonReport: JsonReport) => {
    // Find common root directory
    let trimPath = 0;
    const filepaths = Object.keys(jsonReport.coverageMap);
    if (filepaths.length) {
        const commonRootPath = findCommonPath(filepaths);
        trimPath = commonRootPath.length;
    }

    return Object.entries(jsonReport.coverageMap).reduce<CoverageDetailsMap>(
        (acc, [filename, fileCoverage]) => {
            const normalizedFilename = filename.substr(trimPath);
            acc[normalizedFilename] = {
                filename: normalizedFilename,
                statements: getPercents(
                    standardCoveredCounter('s')(fileCoverage),
                    standardTotalCounter('s')(fileCoverage)
                ),
                functions: getPercents(
                    standardCoveredCounter('f')(fileCoverage),
                    standardTotalCounter('f')(fileCoverage)
                ),
                branches: getPercents(
                    coveredBranchesCounter(fileCoverage),
                    totalBranchesCounter(fileCoverage)
                ),
                lines: getPercents(
                    coveredLinesCounter(fileCoverage),
                    totalLinesCounter(fileCoverage)
                ),
            };
            return acc;
        },
        {}
    );
};
