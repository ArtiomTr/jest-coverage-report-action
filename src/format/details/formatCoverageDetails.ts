import { formatCoverageDetailsPart } from './formatCoverageDetailsPart';
import { getDecreasedCoverage } from './getDecreasedCoverage';
import { getNewFilesCoverage } from './getNewFilesCoverage';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { i18n } from '../../utils/i18n';

export const formatCoverageDetails = (
    headDetails: CoverageDetailsMap,
    baseDetails: CoverageDetailsMap | undefined,
    threshold: number | undefined
): string => {
    const decreasedCoverage = getDecreasedCoverage(headDetails, baseDetails);

    return [
        formatCoverageDetailsPart(
            i18n('newFilesCoverage'),
            getNewFilesCoverage(headDetails, baseDetails),
            undefined,
            threshold
        ),
        formatCoverageDetailsPart(
            i18n('decreasedCoverageFiles'),
            decreasedCoverage.headDetails,
            decreasedCoverage.baseDetails,
            threshold
        ),
    ].join('\n');
};
