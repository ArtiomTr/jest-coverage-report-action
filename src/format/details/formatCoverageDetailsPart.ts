import markdownTable from 'markdown-table';

import { getFileCoverageDetailRow } from './getFileCoverageDetailRow';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { createMarkdownSpoiler } from '../../utils/createMarkdownSpoiler';
import { i18n } from '../../utils/i18n';
import { withExplanation } from '../../utils/withExplanation';

export const formatCoverageDetailsPart = (
    summary: string,
    headDetails: CoverageDetailsMap,
    baseDetails?: CoverageDetailsMap,
    threshold?: number
): string | undefined => {
    const tableContent = Object.keys(headDetails).map((filename) =>
        getFileCoverageDetailRow(
            filename,
            headDetails[filename],
            baseDetails?.[filename],
            threshold
        )
    );

    if (tableContent.length > 0) {
        return createMarkdownSpoiler({
            body: markdownTable(
                [
                    [
                        withExplanation(
                            i18n('status'),
                            i18n('statusExplanation')
                        ),
                        i18n('filename'),
                        i18n('statements'),
                        i18n('branches'),
                        i18n('functions'),
                        i18n('lines'),
                    ],
                    ...tableContent,
                ],
                {
                    align: ['c', 'l', 'l', 'l', 'l', 'l'],
                }
            ),
            summary,
        });
    }

    return undefined;
};
