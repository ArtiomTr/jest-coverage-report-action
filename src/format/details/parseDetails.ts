import { findCommonPath } from './findCommonPath';
import { CoverageDetailsMap } from '../../typings/Coverage';
import { JsonReport } from '../../typings/JsonReport';
import { getFileCoverageMap } from '../../utils/getFileCoverageMap';
import { getPercents } from '../getPercents';

export const parseDetails = (jsonReport: JsonReport) => {
    // Find common root directory
    const filepaths = Object.keys(jsonReport.coverageMap);
    const trimPath = findCommonPath(filepaths).length;

    const fileCoverageMap = getFileCoverageMap(jsonReport);

    return Object.entries(fileCoverageMap).reduce<CoverageDetailsMap>(
        (acc, [filename, coverage]) => {
            const normalizedFilename = filename.substring(trimPath);

            acc[normalizedFilename] = {
                filename: normalizedFilename,
                statements: getPercents(
                    coverage.coveredStatements,
                    coverage.totalStatements
                ),
                branches: getPercents(
                    coverage.coveredBranches,
                    coverage.totalBranches
                ),
                functions: getPercents(
                    coverage.coveredFunctions,
                    coverage.totalFunctions
                ),
                lines: getPercents(coverage.coveredLines, coverage.totalLines),
            };

            return acc;
        },
        {}
    );
};
