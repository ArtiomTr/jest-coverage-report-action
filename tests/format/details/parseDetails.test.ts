import { parseDetails } from '../../../src/format/details/parseDetails';
import { JsonReport } from '../../../src/typings/JsonReport';
import report from '../../mock-data/jsonReport.json';
import report2 from '../../mock-data/jsonReport2.json';

describe('parseDetails', () => {
    it('should match snapshots', () => {
        expect(parseDetails(report)).toMatchSnapshot();
        expect(
            parseDetails((report2 as unknown) as JsonReport)
        ).toMatchSnapshot();
    });
});
