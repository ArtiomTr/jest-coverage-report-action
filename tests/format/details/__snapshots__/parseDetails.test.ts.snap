// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseDetails > should match snapshots 1`] = `
{
  "CheckboxWithLabel.tsx": {
    "branches": 100,
    "filename": "CheckboxWithLabel.tsx",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "calc.ts": {
    "branches": 100,
    "filename": "calc.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "memory.ts": {
    "branches": 100,
    "filename": "memory.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "sub.ts": {
    "branches": 100,
    "filename": "sub.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "sum.js": {
    "branches": 100,
    "filename": "sum.js",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "sum.ts": {
    "branches": 100,
    "filename": "sum.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
}
`;

exports[`parseDetails > should match snapshots 2`] = `
{
  "annotations/createCoverageAnnotations.ts": {
    "branches": 0,
    "filename": "annotations/createCoverageAnnotations.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "annotations/createFailedTestsAnnotations.ts": {
    "branches": 0,
    "filename": "annotations/createFailedTestsAnnotations.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "annotations/isAnnotationEnabled.ts": {
    "branches": 0,
    "filename": "annotations/isAnnotationEnabled.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "constants/REPORT_PATH.ts": {
    "branches": 100,
    "filename": "constants/REPORT_PATH.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "constants/getReportTag.ts": {
    "branches": 100,
    "filename": "constants/getReportTag.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/annotations/formatCoverageAnnotations.ts": {
    "branches": 0,
    "filename": "format/annotations/formatCoverageAnnotations.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "format/annotations/formatFailedTestsAnnotations.ts": {
    "branches": 0,
    "filename": "format/annotations/formatFailedTestsAnnotations.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "format/annotations/getFailedAnnotationsSummary.ts": {
    "branches": 0,
    "filename": "format/annotations/getFailedAnnotationsSummary.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "format/annotations/getFailedTestsAnnotationsBody.ts": {
    "branches": 100,
    "filename": "format/annotations/getFailedTestsAnnotationsBody.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "format/counters.ts": {
    "branches": 71.42857142857143,
    "filename": "format/counters.ts",
    "functions": 100,
    "lines": 96.55172413793103,
    "statements": 97.43589743589743,
  },
  "format/details/findCommonPath.ts": {
    "branches": 100,
    "filename": "format/details/findCommonPath.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/formatCoverageDetails.ts": {
    "branches": 100,
    "filename": "format/details/formatCoverageDetails.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/formatCoverageDetailsPart.ts": {
    "branches": 100,
    "filename": "format/details/formatCoverageDetailsPart.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/getDecreasedCoverage.ts": {
    "branches": 100,
    "filename": "format/details/getDecreasedCoverage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/getFileCoverageDetailRow.ts": {
    "branches": 100,
    "filename": "format/details/getFileCoverageDetailRow.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/getNewFilesCoverage.ts": {
    "branches": 100,
    "filename": "format/details/getNewFilesCoverage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/parseDetails.ts": {
    "branches": 50,
    "filename": "format/details/parseDetails.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/details/shrinkLongPath.ts": {
    "branches": 50,
    "filename": "format/details/shrinkLongPath.ts",
    "functions": 100,
    "lines": 83.33333333333334,
    "statements": 85.71428571428571,
  },
  "format/formatCoverage.ts": {
    "branches": 50,
    "filename": "format/formatCoverage.ts",
    "functions": 100,
    "lines": 85.71428571428571,
    "statements": 87.5,
  },
  "format/formatErrors.ts": {
    "branches": 12.5,
    "filename": "format/formatErrors.ts",
    "functions": 50,
    "lines": 31.25,
    "statements": 35.294117647058826,
  },
  "format/formatRunReport.ts": {
    "branches": 100,
    "filename": "format/formatRunReport.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/getFormattedCoverage.ts": {
    "branches": 100,
    "filename": "format/getFormattedCoverage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/getFormattedFailures.ts": {
    "branches": 80,
    "filename": "format/getFormattedFailures.ts",
    "functions": 100,
    "lines": 91.66666666666666,
    "statements": 93.33333333333333,
  },
  "format/getPercents.ts": {
    "branches": 100,
    "filename": "format/getPercents.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/summary/formatCoverageSummary.ts": {
    "branches": 100,
    "filename": "format/summary/formatCoverageSummary.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/summary/getSummary.ts": {
    "branches": 100,
    "filename": "format/summary/getSummary.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/summary/getTestRunSummary.ts": {
    "branches": 100,
    "filename": "format/summary/getTestRunSummary.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "format/summary/parseSummary.ts": {
    "branches": 100,
    "filename": "format/summary/parseSummary.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "index.ts": {
    "branches": 0,
    "filename": "index.ts",
    "functions": 0,
    "lines": 0,
    "statements": 0,
  },
  "report/fetchPreviousReport.ts": {
    "branches": 100,
    "filename": "report/fetchPreviousReport.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "report/generateCommitReport.ts": {
    "branches": 100,
    "filename": "report/generateCommitReport.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "report/generatePRReport.ts": {
    "branches": 100,
    "filename": "report/generatePRReport.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/collectCoverage.ts": {
    "branches": 100,
    "filename": "stages/collectCoverage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/createReport.ts": {
    "branches": 100,
    "filename": "stages/createReport.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/getCoverage.ts": {
    "branches": 100,
    "filename": "stages/getCoverage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/installDependencies.ts": {
    "branches": 100,
    "filename": "stages/installDependencies.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/parseCoverage.ts": {
    "branches": 100,
    "filename": "stages/parseCoverage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/runTest.ts": {
    "branches": 100,
    "filename": "stages/runTest.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "stages/switchBranch.ts": {
    "branches": 100,
    "filename": "stages/switchBranch.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "typings/Options.ts": {
    "branches": 50,
    "filename": "typings/Options.ts",
    "functions": 100,
    "lines": 96.7741935483871,
    "statements": 97.05882352941177,
  },
  "typings/Report.ts": {
    "branches": 100,
    "filename": "typings/Report.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/DataCollector.ts": {
    "branches": 100,
    "filename": "utils/DataCollector.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/createMarkdownSpoiler.ts": {
    "branches": 100,
    "filename": "utils/createMarkdownSpoiler.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/decimalToString.ts": {
    "branches": 100,
    "filename": "utils/decimalToString.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/formatPercentage.ts": {
    "branches": 100,
    "filename": "utils/formatPercentage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/formatPercentageDelta.ts": {
    "branches": 100,
    "filename": "utils/formatPercentageDelta.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/getConsoleLink.ts": {
    "branches": 100,
    "filename": "utils/getConsoleLink.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/getStatusOfPercents.ts": {
    "branches": 100,
    "filename": "utils/getStatusOfPercents.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/i18n.ts": {
    "branches": 100,
    "filename": "utils/i18n.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/insertArgs.ts": {
    "branches": 100,
    "filename": "utils/insertArgs.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/joinPaths.ts": {
    "branches": 100,
    "filename": "utils/joinPaths.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/runStage.ts": {
    "branches": 100,
    "filename": "utils/runStage.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
  "utils/withExplanation.ts": {
    "branches": 100,
    "filename": "utils/withExplanation.ts",
    "functions": 100,
    "lines": 100,
    "statements": 100,
  },
}
`;
